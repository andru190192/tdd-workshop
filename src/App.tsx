import { useState } from "react";
import reactLogo from "./assets/react.svg";
import "./App.css";
import { add } from "./match";
import { getUserProfile } from "./services/user.service";
import { UserInfoInterface } from "./services/interfaces/user.interface";

function App() {
  const [result, setResult] = useState(0);
  const [user, setUser] = useState<UserInfoInterface | null>(null);

  const hanlderAdd = () => {
    setResult(add([3, 5]));
  };

  const getUserInfo = async () => {
    const userData = await getUserProfile("andru190192");
    setUser(userData);
  };

  return (
    <div className="App">
      <h1 className="text-3xl font-bold underline">Hello world!</h1>
      <h1>3 + 5</h1>
      <h1>{result}</h1>
      <div className="card">
        <button onClick={hanlderAdd}>Sumar</button>
      </div>
      <div className="card">
        <button onClick={getUserInfo}>Ver Perfil GitHub</button>
      </div>
      {user && (
        <div>
          <img
            className="inline-block h-6 w-6 rounded-full ring-2 ring-white"
            src={user.avatar_url}
            alt=""
          />
          <p>{user.name}</p>
          <p>{user.location}</p>
        </div>
      )}
    </div>
  );
}

export default App;
