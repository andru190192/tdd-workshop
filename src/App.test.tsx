import React from "react";
import { act, render, fireEvent, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import App from "./App";
import * as UserService from "./services/user.service";

describe("App Component", () => {
  it("should render correctly", () => {
    const { container } = render(<App />);

    expect(container).toMatchSnapshot();
    expect(screen.getByText("3 + 5")).toBeInTheDocument();
    expect(screen.getByText("Sumar").closest("button")).toBeInTheDocument();
    expect(screen.getByText("Ver Perfil GitHub")).toBeInTheDocument();
  });

  it("should display the sum result when click on add button", () => {
    render(<App />);

    expect(screen.getByText("3 + 5")).toBeInTheDocument();

    const addBtn = screen.getByText("Sumar") as HTMLButtonElement;

    fireEvent.click(addBtn);

    expect(screen.getByText("8")).toBeInTheDocument();
  });

  it("should display the user info when click on github profile button", async () => {
    // Arrange
    vi.spyOn(UserService, "getUserProfile").mockReturnValue(
      Promise.resolve({
        name: "Andres Guachisaca",
        avatar_url: "https://avatars.githubusercontent.com/u/8331153?v=4",
        location: "Quito",
      })
    );

    // Act
    render(<App />);

    const githubProfileBtn = screen.getByText(
      "Ver Perfil GitHub"
    ) as HTMLButtonElement;

    await act(async () => {
      fireEvent.click(githubProfileBtn);
    });

    // Asserts
    expect(screen.getByText("Andres Guachisaca")).toBeInTheDocument();
    expect(screen.getByRole("img")).toHaveAttribute(
      "src",
      "https://avatars.githubusercontent.com/u/8331153?v=4"
    );
    expect(screen.getByText("Quito")).toBeInTheDocument();
  });
});
