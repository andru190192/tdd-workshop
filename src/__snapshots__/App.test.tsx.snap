// Vitest Snapshot v1

exports[`App Component > should render correctly 1`] = `
<div>
  <div
    class="App"
  >
    <h1
      class="text-3xl font-bold underline"
    >
      Hello world!
    </h1>
    <h1>
      3 + 5
    </h1>
    <h1>
      0
    </h1>
    <div
      class="card"
    >
      <button>
        Sumar
      </button>
    </div>
    <div
      class="card"
    >
      <button>
        Ver Perfil GitHub
      </button>
    </div>
  </div>
</div>
`;
